### PR TITLE
Add support for truncate to lower for streaming writes and fix promotion to generation backed Inode in Write and Truncate Flow.

### PR DESCRIPTION
**Pull request by meet2mky**
### Description
Add support for truncate to lower size than current file size during streaming writes. We are going to finalise the file and fallback to stage disk. This PR also fixes the promotion to generation backed Inodes for both out of order write flow and truncate to lower size scenario.

### Link to the issue in case of a bug fix.
b/421367385

### Testing details
1. Manual - Done
2. Unit tests - Updated.
3. Integration tests - Ran as part of presubmit.

### CODE REVIEW

File: internal/fs/inode/file.go [hunk 1]
```diff
@@ -934,7 +934,22 @@ func (f *FileInode) Truncate(
 	size int64) (err error) {
 
 	if f.bwh != nil {
-		return f.bwh.Truncate(size)
+		err = f.bwh.Truncate(size)
+		if err == nil {
+			return
+		}
+	}
+	if err != nil && !errors.Is(err, bufferedwrites.ErrTruncateSizeLessThanFileSize) {
```
Comment (ashmeenkaur) [inline internal/fs/inode/file.go:942]: We can remove the redundant err != nil check, as we would have already returned if err were nil. This check should also only occur when bwh != nil. Perhaps we should refactor this into a separate method.

Following the Go style guide for error handling, I recommend ordering the checks as follows:
- errors.Is(err, bufferedwrites.ErrTruncateSizeLessThanFileSize)
- err != nil
- The normal flow (not indented)

File: internal/fs/inode/file.go [hunk 2]
```diff
@@ -934,7 +934,22 @@ func (f *FileInode) Truncate(
...
+		return fmt.Errorf("got unexpected error from bwh.Truncate(): %w", err)
+	}
+	// If truncate size is less than total size, flush and fall back to temp file.
+	if errors.Is(err, bufferedwrites.ErrTruncateSizeLessThanFileSize) {
+		logger.Infof("Falling back to staged writes on disk for file %s (inode %d) due to err: %v.", f.Name(), f.ID(), err.Error())
+		// Finalize the object.
+		err = f.flushUsingBufferedWriteHandler()
```
Comment (ashmeenkaur) [inline internal/fs/inode/file.go:949]: Don't we have to promote the inode to generation backed inode if it was local? Seems like this is missing in out of order write flow as well.

File: internal/bufferedwrites/buffered_write_handler.go [hunk 3]
```diff
@@ -260,7 +261,9 @@ func (wh *bufferedWriteHandlerImpl) SetMtime(mtime time.Time) {
 
 func (wh *bufferedWriteHandlerImpl) Truncate(size int64) error {
 	if size < wh.totalSize {
-		return fmt.Errorf("cannot truncate to lesser size when upload is in progress")
+		logger.Errorf("BufferedWriteHandler.TruncateToLowerSizeError for object: %s, file size: %d, truncate size: %d",
```
[the remark this answers is not in the document]
Comment (meet2mky) [reply internal/bufferedwrites/buffered_write_handler.go]: Removed duplicate log from here and added warning log in file.go.

File: internal/fs/inode/file.go [hunk 4]
```diff
@@ -934,7 +934,22 @@ func (f *FileInode) Truncate(
...
+		if err == nil {
+			return
+		}
+	}
+	if err != nil && !errors.Is(err, bufferedwrites.ErrTruncateSizeLessThanFileSize) {
+		return fmt.Errorf("got unexpected error from bwh.Truncate(): %w", err)
+	}
+	// If truncate size is less than total size, flush and fall back to temp file.
+	if errors.Is(err, bufferedwrites.ErrTruncateSizeLessThanFileSize) {
+		logger.Infof("Falling back to staged writes on disk for file %s (inode %d) due to err: %v.", f.Name(), f.ID(), err.Error())
```
[the remark this answers is not in the document]
Comment (meet2mky) [reply internal/fs/inode/file.go]: Logging warning log here only, removed from buffered_write_handler.go

File: internal/fs/inode/file.go (hunk 1 above)
Comment (meet2mky) [reply internal/fs/inode/file.go]: I have fixed the flow as per review. I think refactoring we can do for both the Truncate and Write in a separate PR as both have similar error handling now.

File: internal/fs/inode/file.go (hunk 2 above)
Comment (meet2mky) [reply internal/fs/inode/file.go:940]: Made all the changes along with truncate method refactoring, integration tests and fs_tests. PTAL

File: internal/fs/fs.go [hunk 5]
```diff
@@ -2654,7 +2659,7 @@ func (fs *fileSystem) WriteFile(
 		fs.mu.Unlock()
 
 		//TODO: Initialize BWH before invoking write()
-		if err := fh.Write(ctx, op.Data, op.Offset); err != nil {
+		if _, err := fh.Write(ctx, op.Data, op.Offset); err != nil {
```
Comment (meet2mky) [inline internal/fs/fs.go:2662]: This change is under experimental flag and work in progress so I haven't changed it's code for this.

File: internal/fs/handle/file.go [hunk 6]
```diff
@@ -161,7 +161,7 @@ func (fh *FileHandle) Read(ctx context.Context, dst []byte, offset int64, sequen
 // unfinalized objects in zonal buckets, streaming writes is used.
 // Note that the writes are still done at the inode level.
 // LOCKS_EXCLUDED(fh.inode)
-func (fh *FileHandle) Write(ctx context.Context, data []byte, offset int64) error {
+func (fh *FileHandle) Write(ctx context.Context, data []byte, offset int64) (bool, error) {
```
Comment (meet2mky) [inline internal/fs/handle/file.go:164]: This change is under experimental flag and work in progress so I haven't changed it's code for this. Will talk to anushka to align on this.

File: internal/fs/handle/file_test.go [hunk 7]
```diff
@@ -116,7 +116,7 @@ func TestFileHandleWrite(t *testing.T) {
 	ctx := context.Background()
 	data := []byte("hello")
 
-	err := fh.Write(ctx, data, 0)
+	_, err := fh.Write(ctx, data, 0)
```
Comment (meet2mky) [inline internal/fs/handle/file_test.go:119]: This change is under experimental flag and work in progress so I haven't changed it's code for this. Will talk to anushka to align on this.

File: internal/fs/inode/file_test.go [hunk 8]
```diff
@@ -1052,22 +1070,20 @@ func (t *FileTest) TestTruncateDownwardWhenStreamingWritesAreEnabled() {
...
-				assert.Equal(t.T(), int64(7), t.in.bwh.WriteFileInfo().TotalSize)
-				// Fetch the attributes and check if the file size reflects the write.
-				attrs, err := t.in.Attributes(t.ctx)
-				require.NoError(t.T(), err)
-				assert.Equal(t.T(), uint64(7), attrs.Size)
-			}
-			t.createBufferedWriteHandler(!tc.performWrite)
-
-			err = t.in.Truncate(t.ctx, tc.truncateSize)
+			t.createBufferedWriteHandler(true)
```
Comment (ashmeenkaur) [inline internal/fs/inode/file_test.go:1083]: Can we keep test case for how it was earlier - covering the edge case when truncate is done before initializing bwh?

can we also add a scenario for negative truncate size which should result in an error (as a separate test)

File: internal/fs/inode/file_test.go [hunk 9]
```diff
@@ -1052,22 +1070,20 @@ func (t *FileTest) TestTruncateDownwardWhenStreamingWritesAreEnabled() {
...
+			attrs, err = t.in.Attributes(t.ctx)
+			require.NoError(t.T(), err)
+			assert.Equal(t.T(), uint64(7), attrs.Size)
+			gcsSynced, err = t.in.Truncate(t.ctx, tc.truncateSize)
 
-			require.Error(t.T(), err)
-			assert.ErrorContains(t.T(), err, "cannot truncate")
+			require.NoError(t.T(), err)
+			assert.True(t.T(), gcsSynced)
+			t.createBufferedWriteHandler(false)
```
Comment (ashmeenkaur) [inline internal/fs/inode/file_test.go:1096]: why is this required?

File: tools/integration_tests/local_file/stat_file.go [hunk 10]
```diff
@@ -56,12 +56,13 @@ func (t *CommonLocalFileTestSuite) TestStatOnLocalFileWithConflictingFileNameSuf
...
-func (t *localFileTestSuite) TestTruncateLocalFileToSmallerSize() {
+func (t *CommonLocalFileTestSuite) TestTruncateLocalFileToSmallerSize() {
 	testDirPath = setup.SetupTestDirectory(testDirName)
 	// Create a local file.
-	filePath, fh := CreateLocalFileInTestDir(ctx, storageClient, testDirPath, FileName1, t.T())
+	fileName := FileName1 + setup.GenerateRandomString(5)
+	filePath, fh := CreateLocalFileInTestDir(ctx, storageClient, testDirPath, fileName, t.T())
 	// Writing contents to local file .
-	WritingToLocalFileShouldNotWriteToGCS(ctx, storageClient, fh, testDirName, FileName1, t.T())
+	operations.WriteWithoutClose(fh, FileContents, t.T())
```
Comment (ashmeenkaur) [inline tools/integration_tests/local_file/stat_file.go:65]: This change is because of ZB, is it?

File: tools/integration_tests/streaming_writes/truncate_file_test.go [hunk 11]
```diff
@@ -84,51 +86,116 @@ func (t *StreamingWritesSuite) TestWriteAfterTruncate() {
...
+			operations.WriteWithoutClose(t.f1, tc.writeContent, t.T())
+
+			operations.VerifyStatFile(t.filePath, int64(len(tc.finalContent)), FilePerms, t.T())
+			// Close the file and validate that the file is created on GCS.
+			CloseFileAndValidateContentFromGCS(ctx, storageClient, t.f1, testDirName, t.fileName, tc.finalContent, t.T())
+		})
+	}
+}
 
+func (t *StreamingWritesSuite) TestTruncateToLowerSyncsFileToGcsForStreamingWrites() {
```
Comment (ashmeenkaur) [inline tools/integration_tests/streaming_writes/truncate_file_test.go]: This overlaps with TestWriteAndTruncate, can we revert the changes in TestWriteAndTruncate and keep the 2 scenarios different like how it was earlier.

File: tools/integration_tests/streaming_writes/truncate_file_test.go [hunk 12]
```diff
@@ -84,51 +86,116 @@ func (t *StreamingWritesSuite) TestWriteAfterTruncate() {
...
+	require.NoError(t.T(), err)
+	operations.VerifyStatFile(t.filePath, 3, FilePerms, t.T())
+	if !t.fallbackToDiskCase {
+		ValidateObjectContentsFromGCS(ctx, storageClient, testDirName, t.fileName, "foobar", t.T())
+	}
+	CloseFileAndValidateContentFromGCS(ctx, storageClient, t.f1, testDirName, t.fileName, "foo", t.T())
 }
 
-func (t *StreamingWritesSuite) TestTruncateToLowerSizeAfterWrite() {
+func (t *StreamingWritesSuite) TestTruncateToLowerSyncsFileToGcsAndDeletingFileDeletsFileFromGcs() {
```
Comment (ashmeenkaur) [inline tools/integration_tests/streaming_writes/truncate_file_test.go]: NIT:
```suggestion
func (t *StreamingWritesSuite) TestTruncateDownAndDeleteFile() {
```

File: tools/integration_tests/local_file/stat_file.go (hunk 10 above)
Comment (meet2mky) [reply tools/integration_tests/local_file/stat_file.go:65]: yes, writing to local file can create file on GCS in case of ZB for streaming writes.

File: internal/fs/inode/file_test.go (hunk 8 above)
Comment (meet2mky) [reply internal/fs/inode/file_test.go:1083]: Truncate is never done before trying to initialize bwh as that logic is at fs layer. This test case before also we have initialized bwh before every inode.Truncate call to ensure similar behaviour of fs layer.
Added negative size truncate tests to both the streaming writes file and current file which should cover all the cases.

File: internal/fs/inode/file_test.go (hunk 9 above)
Comment (meet2mky) [reply internal/fs/inode/file_test.go:1096]: This is required to ensure that after truncate to size lower than current size or even zero size. We won't reinitialise bwh for it.

File: tools/integration_tests/streaming_writes/truncate_file_test.go (hunk 12 above)
Comment (meet2mky) [reply tools/integration_tests/streaming_writes/truncate_file_test.go]: Done.

File: tools/integration_tests/streaming_writes/truncate_file_test.go (hunk 11 above)
Comment (meet2mky) [reply tools/integration_tests/streaming_writes/truncate_file_test.go]: Removed this test as it is covered as part of other test. I am keeping the TestWriteAndTruncate as current one is more readable and covers both scenario for truncate to lower and upper.

**Comment by ashmeenkaur** [comment]
@anushka567 FYI

**Review by ashmeenkaur — APPROVED** [review]

**Comment by anushka567** [comment]
Noted. I will take care of the required changes when I am working on this. I have created a [bug](b/422075967) for the same to track it.
